metadata_writer_nne.c

```diff
@@ -94,132 +309,725 @@ static uint8_t md_nne_handle_gps_event(struct md_writer_nne *mwn,
+                          mie->imsi_mccmnc, descr->idx, descr->key,
+                          modem->metadata[descr->idx].value.u.v_uint8);
+
+        // Generate json event
+        struct nne_message msg;
+        msg.type = NNE_MESSAGE_TYPE_EVENT;
+        msg.tstamp = mie->tstamp;
+        msg.node = mwn->node_id;
+        msg.mccmnc = mie->imsi_mccmnc;
+        msg.key = descr->key;
+        msg.value = value;
+        msg.extra = NULL;
+        msg.source = source;
+        msg.delta = 0;
+
+        md_nne_send_message(mwn, &msg);
+
+        // if mode has changed nullify mode dependend metadata
+        if (descr->idx == NNE_IDX_MODE) {
+            for(i = 0; i <= NNE_IDX_MAX; i++) {
+                if (descr->mode_dependent) {
+                    if (modem->metadata[i].value.type == NNE_TYPE_STRING) {
+                        free(modem->metadata[i].value.u.v_str);
+                    }
+                    modem->metadata[i].value.type = NNE_TYPE_NULL;
+
+                    META_PRINT_SYSLOG(mwn->parent, LOG_ERR,
+                            "NNE writer: mccmnc %d: %d,%s nullified\n",
+                            mie->imsi_mccmnc, descr->idx, descr->key);
+                }
+            }
+        }
+    }
+    else {
+        if (value.type == NNE_TYPE_STRING) {
+            free(value.u.v_str);
+        }
+    }
+
+    modem->metadata[descr->idx].tstamp = mie->tstamp;
+}
+
+static void md_nne_generate_bins1min(struct md_writer_nne *mwn,
+                                        uint64_t tstamp)
+{
+    struct nne_modem *modem = NULL;
+    struct nne_metadata_descr *descr;
+    struct nne_message msg;
+    int i;
+
+    LIST_FOREACH(modem, &(mwn->modem_list), entries)
+    {
+        // On first event after new minute has begun
+        // generate bins1min entries for all metadata values
+        if ((tstamp / 60) <= (modem->tstamp / 60))
+            continue;
+
+        for (i = 0; i < NNE_METADATA_DESCR_LEN; i++)
+        {
+            descr = &(NNE_METADATA_DESCR[i]);
+
+            msg.type = NNE_MESSAGE_TYPE_BINS1MIN;
+            msg.tstamp = (tstamp / 60) * 60;
+            msg.node = mwn->node_id;
+            msg.mccmnc = modem->mccmnc;
+            msg.key = descr->key;
+            msg.value = modem->metadata[descr->idx].value;
+            msg.extra = NULL;
+            msg.source = NNE_MESSAGE_SOURCE_QUERY;
+            msg.delta = msg.tstamp - modem->metadata[descr->idx].tstamp;
+
+            md_nne_send_message(mwn, &msg);
+        }
+    }
+}
+
+static void md_nne_check_removed_modems(struct md_writer_nne *mwn,
+                                        uint64_t tstamp)
+{
+    struct nne_modem *modem = NULL, *rm_modem = NULL;
+    struct nne_message msg;
+
+    modem = mwn->modem_list.lh_first;
+    while(modem != NULL)
+    {
+        // check if modem has been removed
+        rm_modem = NULL;
+        if (modem->tstamp < tstamp - 40)
```

Comment:
Is 10 second slack enough? I don't know if protocol used for nornet is reliable enough, so what if we loose a packet?